File: app/src/main/java/org/antrack/app/ui/fragments/ShellFragment.java
package org.antrack.app.ui.fragments;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;

import org.antrack.app.libs.Files;
import org.antrack.app.libs.Keyboard;
import org.antrack.app.libs.L;
import org.antrack.app.ui.State;
import org.antrack.app.ui.U;

import java.io.IOException;

import app.R;

public class ShellFragment extends BaseFragment {
    private final String TAG = "ShellFragment";

    Context context;

    EditText editText;
    TextView textView;
    TextView ps1;

    String cmdOut;
    String cmdCmd;

    boolean progress = false;
    Thread progressThread;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Otherwise GetActivity() return null after orientation change
        setRetainInstance(true);

        if (!Mod.check(Mod.CMD)) {
            showNoModule(Mod.CMD);
            return null;
        }

        cmdOut = Mod.getFile(Mod.CMD);
        cmdCmd = Mod.getCommand(Mod.CMD);

        context = getActivity().getApplicationContext();

        View view = inflater.inflate(R.layout.fragment_shell, container, false);
        view.setAlpha(0);
        view.animate().alpha(1);

        // Show keyboard when user clicks on any place
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Keyboard.show(getActivity());
            }
        });

        ps1 = (TextView) view.findViewById(R.id.fragment_shell_ps1);
        String text = State.device.getName() + "$ ";
        ps1.setText(text);

        editText = (EditText) view.findViewById(R.id.fragment_shell_edittext);
        editText.requestFocus();

        textView = (TextView) view.findViewById(R.id.fragment_shell_textview);
        textView.setMovementMethod(new ScrollingMovementMethod());
        textView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Keyboard.show(getActivity());
            }
        });

        editText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                boolean handled = false;
                if (actionId == EditorInfo.IME_ACTION_SEND) {
                    sendCommand(editText.getText().toString());

                    editText.setText("");
                    textView.setText("");

                    startProgress();

                    handled = true;
                }
                return handled;
            }
        });

        return view;
    }

    private void sendCommand(String cmd) {
        U.runCommandAsync(cmdCmd + " " + cmd);
    }

    private void addText(final String text) {
        Activity activity = getActivity();

        if (activity == null) return;
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                textView.append(text);
            }
        });
    }

    private void startProgress() {
        if (State.device.isMain())
            return;

        stopProgress();

        progressThread = new Thread(new Runnable() {
            @Override
            public void run() {
                int seconds = 0;

                while (true) {
                    addText(".");

                    if (seconds == 60) {
                        addText(" :(");
                    }

                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        return;
                    }
                    seconds += 1;
                }
            }
        });
        progressThread.start();
    }

    private void stopProgress() {
        if (State.device.isMain())
            return;

        if (progressThread != null && progressThread.isAlive()) {
            progressThread.interrupt();
        }
    }

    @Override
    public void onFileUpdate() {
        try {
            final String out = Files.readTextFile(U.getLocalPath(cmdOut));

            getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    stopProgress();
                    textView.setAlpha(0);
                    // Remove time stamp
                    textView.setText(out.substring(out.indexOf('\n')+1));
                    textView.animate().alpha(1);
                }
            });
        } catch (IOException e) {
            L.e(TAG, "Can't read cmdout: " + e);
        }
    }

    @Override
    public String getWatchFile() {
        return cmdOut;
    }
}

File: app/src/main/java/org/antrack/app/ui/fragments/ShellFragment.kt
@file:Suppress("OVERRIDE_DEPRECATION", "DEPRECATION")

package org.antrack.app.ui.fragments

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.TextView
import app.R
import org.antrack.app.Env
import org.antrack.app.RESULT_FILE
import org.antrack.app.functions.*
import org.antrack.app.modules.ModuleInterface
import org.antrack.app.modules.Modules
import org.antrack.app.watcher.FileWatcher
import org.antrack.app.watcher.IWatcherCallback
import java.io.File

class ShellFragment : BaseFragment() {
    private lateinit var resultView: TextView

    private var lastCommand = ""

    inner class FragmentCallback : IWatcherCallback {
        override val watchFile = RESULT_FILE
        override fun onFileUpdated(path: String) {
            onResultFileUpdated()
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Otherwise GetActivity() return null after orientation change
        retainInstance = true

        val main = inflater.inflate(R.layout.fragment_shell, container, false)
        val ps1 = main.findViewById(R.id.fragment_shell_ps1) as TextView
        val inputField = main.findViewById(R.id.fragment_shell_edittext) as EditText
        resultView = main.findViewById(R.id.fragment_shell_textview) as TextView

        ps1.text = Env.deviceName + "$ "
        inputField.requestFocus()

        setListeners(main, resultView, inputField)

        return main
    }

    override fun onStart() {
        super.onStart()
        FileWatcher.addCallback(className, FragmentCallback())
    }

    override fun onStop() {
        super.onStop()
        resultView.hideKeyboard()
        FileWatcher.removeCallback(className)
    }

    private fun setListeners(
        mainView: View,
        resultView: TextView,
        inputField: EditText,
    ) {
        mainView.setOnClickListener {
            inputField.focusAndShowKeyboard()
        }

        resultView.setOnClickListener {
            inputField.focusAndShowKeyboard()
        }

        inputField.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                onEnterPressed(inputField)
            }
            true
        }
    }

    private fun onEnterPressed(inputField: EditText) {
        lastCommand = inputField.text.toString()
        inputField.setText("")

        if (lastCommand.isNotBlank()) {
            runCommandAsync(lastCommand)
            resultView.setText(R.string.loading)
        }
    }

    private fun onResultFileUpdated() = async {
        try {
            val module = Modules.get().getOrElse(lastCommand.split(" ")[0]) { null }

            val out = when {
                module != null && module.resultType() == "txt" -> readModuleOutFile(module)
                else -> readResultFile()
            }

            insertOutInResultViewOnUiThread(out)
        } catch (e: Exception) {
            logE(className, "Can't read result file: $e")
            activity.toast(e.message.toString())
        }
    }

    private fun insertOutInResultViewOnUiThread(out: String) {
        activity.runOnUiThread {
            resultView.text = out
            resultView.fadeIn()
        }
    }

    private fun readResultFile() = File(Env.resultFilePath).readText()

    private fun readModuleOutFile(module: ModuleInterface): String {
        val resultFile = File(Env.mainDirPath + module.result())

        return if (resultFile.isDirectory) {
            resultFile.listFiles()?.joinToString("\n") { it.readText() } ?: "error"
        } else {
            resultFile.readText()
        }
    }
}